### PR TITLE
Automate Type URI generation

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

When Types are made through the workspace, we don't want users having to manually choose appropriate (and valid) URIs.
This PR abstracts that way, automating the process of the URIs being generated. It makes a large number of drivebys as well, to documentation, testing strategy, etc. (These were somewhat necessary in development, apologies they weren't separated out better).

There were also other concerns like making sure URIs are valid and that the workspace is capable of serving them.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202848989198299/f) _(internal)_

## 🔍 What does this change?

- Adds the `DomainValidator` struct in the Graph which uses a configured Regex to validate type URIs.
- Changes the signature of all `update*` methods on the Graph API to take a schema without an $id, and now an additional parameter `typeToUpdate`
  - Changes the implementation of the entry point of those methods to increment the `typeToUpdate` URI, and to inject it into the schema before reconstructing a full type
- Adds the `VAR_UPDATE_*` json schemas to use in the Graph API methods

- Completely reworks how seed data is created/ensured on the Workspace. A possibly too-magic implementation of async function initializers is used to handle initialization logic and dependencies.
- Modifies all resolvers to handle the new signatures
- `create*` resolvers for communicating with the frontend now take `Omit<type, "$id">` and generate a type at that time using the given accountId (which requires finding the appropriate shortname)

- Rewrites all relevant tests to create types on appropriate domains

## 📜 Does this require a change to the docs?

- The generated documentation has been adjusted. Some additional improvements have been added to the READMEs.
- We almost certainly need to improve docs around running tests, ensuring a clean DB, etc.

## ⚠️ Known issues
- There is a race condition inside the Graph where we check for an existing URI. This means we have to run integration tests sequentially
- Many more, see comment threads @Alfred-Mountfield has opened. Ensure these are resolved before merge.

## 🐾 Next steps

- Also many, see comment threads and relevant TODOs in the codebase (asana tasks have been included)

## 🛡 What tests cover this?

- Rust unit tests
- Rust integration tests
- HTTP tests for the Graph REST API
- Workspace integration tests for:
  -  "knowledge"
  - "ontology"

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2. Try running the Graph tests according to the readme.
3. Ensure you take down the docker deployment and clear all volumes.
4. Run `external-services up --build` in the repo root
5. Inside `packages/hash/integration` in another terminal run
  a.  `LOG_LEVEL="debug" yarn jest "ontology" --maxWorkers 1`  
  b.  `LOG_LEVEL="debug" yarn jest "knowledge" --maxWorkers 1`
6. Also try `yarn dev` and check that login works, and that the `/type-editor` route is not broken.

### CODE REVIEW

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bin/hash_graph/src/args.rs:40]: Are there any env files I need to update to override this default?

File: packages/graph/hash_graph/bin/hash_graph/src/main.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bin/hash_graph/src/main.rs:37]: I'm not really sure what to do about these, and I'm feeling a little burned out from this task now.

File: packages/graph/hash_graph/lib/graph/src/api/rest/json_schemas/update_data_type.json
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/json_schemas/update_data_type.json:1]: These are literally just copy pastes of their non `update_` equivalents, with "$id" removed.

This is obviously less than ideal but I don't know how we want to address it right now without a tonne of overhead, and we might not be serving these meta-schemas statically for long anyway.

File: packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts:22]: Why are we allowing entity type to be missing? Just for convenience to not have to pass the ID?

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/user.model.ts:37]: What should we do about this in the interim after discussion today.

Just attach an asana task about dealing with accountIds?

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/user.model.ts:71]: We don't really do this anywhere else, I'm happy to delete it (especially because data isn't always there) but this is generally the problem I've been encountering where panics aren't giving the right information.

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/user.model.ts:267]: Since writing it I've come to learn that shortnames can be missing during User creation? I think we need to revisit this, as shortnames should definitely be an invariant outside of the creation flow. We shouldn't have to deal with conditionals all over the place and possibly have one being missing.

File: packages/hash/api/src/model/ontology/data-type.model.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/ontology/data-type.model.ts:51]: The workspace account doesn't have an associated user, so I had to grab it from the env-var because `getUserByAccountId` fails. Obviously this is less than ideal.

File: packages/hash/api/src/model/util.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/util.ts:78]: We're passing in magic strings, and this seems to just avoid us passing an `"entity-type"` string rather than `"entityType"`... I don't really see value in these so I'm proposing we just swap the top type to `: "entity-type" | "property-type" | ..`

File: packages/hash/backend-utils/src/system.ts
Comment (Alfred-Mountfield) [inline packages/hash/backend-utils/src/system.ts:3]: Inline with previous comment about a workspace account not having an associated User

File: packages/hash/integration/src/tests/model/ontology/link-type.model.test.ts
Comment (Alfred-Mountfield) [inline packages/hash/integration/src/tests/model/ontology/link-type.model.test.ts:34]: These somewhat weird shortnames are because I'm not sure if we can use `kebab-case`, capitals, numbers, etc.

File: packages/hash/integration/src/tests/util.ts
Comment (Alfred-Mountfield) [inline packages/hash/integration/src/tests/util.ts:99]: This is kind of gross but I was running into "CLASH" errors from Kratos otherwise. I'm still not entirely sure why if I'm honest.

Comment (Alfred-Mountfield) [reply packages/hash/integration/src/tests/util.ts:99]: At the very least it should give us quite a few test runs before the DB becomes too populated with previous data.

Again, this is a terrible patch and we really need to figure out clean testing environments.

File: packages/hash/frontend/src/pages/type-editor/index.page.tsx
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/type-editor/index.page.tsx:107]: I didn't bother to replace the way to update the URI as the type-editor is being worked upon now. The create button still works for first click so I think it's okay?

File: packages/graph/hash_graph/bin/hash_graph/src/main.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:37]: I assume that we want to remove this function somewhat soon, so I think it's fine to not validate them for now.

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:37]: I'm somewhat inclined to agree but I'm not sure it'll be super soon. I think these will only really be removed when we they are hosted somewhere else, and when we have to handle the logic of storing metadata records about types that live somewhere else (as opposed to creating a new type)

I _think_ it's probably not a big deal, I just worry we'll ignore this for a while and then there'll be some logic that involves crawling through links or exploring definitions and we'll try and resolve these and it'll break.

File: packages/hash/api/src/model/util.ts
Comment (thehabbos007) [inline packages/hash/api/src/model/util.ts:291]:
```suggestion
 * Helper method for generating an entity type schema for the Graph API.
```

**Review by thehabbos007 — COMMENTED** [review]
Thanks for taking this on, Alfie! I've looked through most of it here :)

File: packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts:22]: Sorry yes that's what I meant, have edited. I don't think I understand though. Why do we allow `accountId` to be _missing_, surely that wouldn't create a type on another account.

I guess I'm saying why don't we just always require an accountId, how convenient is it really to sometimes let it be undefined and to grab it from the current user?

File: packages/hash/api/src/model/util.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/util.ts:78]: I'm just saying we're passing strings, so there's no need to stick to camelCase. They aren't variables haha

File: packages/hash/integration/src/tests/model/ontology/link-type.model.test.ts
Comment (Alfred-Mountfield) [reply packages/hash/integration/src/tests/model/ontology/link-type.model.test.ts]: Have updated this, kebab case is apparently allowed. I've also updated the default regex to reflect

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:97]: Dang I thought I'd caught all of them. Thanks :D

File: packages/graph/hash_graph/lib/graph/src/api/rest/link_type.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/api/rest/link_type.rs:213]: I'm not very happy that we force an increment of one. If I remember correctly, we decided to not do it as either (a) people could choose to not publish a version, so they have a few revisions locally and only publish the latest, or (b) some revisions are also hosted elsewhere (with for whatever reason has the same base URI).
I don't want to block merging on this as this is a scenario we don't need to cover right now, but it would be great if we have a task covering this.

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/link_type.rs:213]: Some good thoughts, I think they're not issues for a number of reasons:
1. We've talked about "unpublished" versions as being a UI nicety, and as expressed I really don't think they should live as _**types**_ within the graph. We've floated other options such as them being stored as entities linked to the relevant blocks, etc. I bring this up because even though we haven't figured out the correct approach for draft types, etc., I think we can rule out storing them as native types. It would greatly complicate the table schemas, validation logic, querying implementations, etc. So I think your point (a) is a non-issue.
1. There are two elements to your statement and I think it's worth clarifying them
    1. Types can be hosted elsewhere - This is correct, but it's worth noting that things that are hosted elsewhere will not live in the same tables or interfaces as types owned by the graph. We will persist metadata records (and cache copies) of the types that live elsewhere, but that is it. They **must** be managed differently because.
       - We cannot serve them, their "$id" will be some other domain we do not own
       - We know nothing about their ownership, the person who owns their domain, or owns the type in their system, may not have a HASH account, and we certainly don't have write access
   1. _Some revisions_ - I apologise if I'm being pedantic but there's a subtle implication here that's incorrect in my mind. A type is uniquely determined by its base domain. We do not have a way to express a change of base domain, and I do not think we ever will, (because there's no way to validate that the owner of the base domain is the same as that who owns the new one, etc.). Thus, there will never be a situation where some revisions of a type live on `example.com` and then new revisions of the type live at a domain hosted by the graph. If a type is hosted by the graph, then its full life cycle will be at the Graph's domain. And we will be good BP citizens and ensure that updates are incremental, because it plays nicer with tooling that tries to find the next and previous version, etc.

File: packages/hash/api/src/model/ontology/data-type.model.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/ontology/data-type.model.ts:47]: Indeed. I think there's _some_ benefit to not having a type alias, as I could see us finally fixing it and then leaving the type alias in place. Whereas here we'd be forced to update all of these to just point to `DataType`. Rather than ending up with something like `type DataType = DataType` and then inconsistently using which.

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/user.model.ts:72]: Have responded to that :) Feel free to resolve this if you're happy with the suggestion

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/user.model.ts]: Have updated the todo

File: packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts:22]: I guess it boils down to the frontend not needing to each for the current user, and letting the Graph QL context/session implicitly pass it. I am not sure how inconvenient it would be to require an accout ID for the frontend

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts:22]: I'll get rid of the todo because I think we'll revisit anyway as we consolidate the accountId stuff.

File: packages/hash/api/src/model/ontology/data-type.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/ontology/data-type.model.ts:83]: Added task here https://app.asana.com/0/1202805690238892/1202945471553579/f. Completely fair to not do it here

File: packages/hash/api/src/model/ontology/data-type.model.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/ontology/data-type.model.ts:51]: Not sure what we want to do for this PR, I don't want to leave the TODO as is

File: packages/graph/README.md
Comment (TimDiekmann) [inline packages/graph/README.md:29]: This renders weirdly
```suggestion
### Logging configuration

```

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/user.model.ts]: Thanks, that seems good!

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs:56]: Do we want to return an error on invalid regexes?

File: packages/hash/api/src/model/util.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/util.ts]: I see the change is applied. Reasonable to me as we're just using it for `generateSchemaUri` :+1:

File: packages/hash/api/src/model/knowledge/user.model.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/user.model.ts:72]: Yeah I didn't want to throw in a todo in this particular case. Thanks!

File: packages/hash/integration/src/tests/util.ts
Comment (Alfred-Mountfield) [reply packages/hash/integration/src/tests/util.ts:99]: cc @TimDiekmann and @thehabbos007 just in case you guys have ideas about cleaner testing environments. (Not asking for a response to this comment thread, just flagging for the next time we discuss it :) )

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/entity_type/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/entity_type/mod.rs:79]:
```suggestion
                    "could not insert references for entity type: {}",
```

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/property_type/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/property_type/mod.rs:41]:
```suggestion
                    "could not insert references for property type: {}",
```

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/property_type/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/property_type/mod.rs:83]:
```suggestion
                    "could not insert references for property type: {}",
```

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs]: Thank you!

File: packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/ontology/entity-type.ts:28]: Ah yes, some of the catches are in the model classes now, fair. You're right that it's unwieldy, but I guess that's how it goes with exceptions that don't need to be added to any signature. Hopefully we can take care of this as we figure out a unified error response object from the API and can get rid of relying on HTTP response codes

File: packages/hash/api/src/model/ontology/data-type.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/ontology/data-type.model.ts:47]: Very fair, I don't think it's worth doing in this case as we do get type information from `DataType`, so at least `Pick` would fail if it's wrong :+1:

File: packages/hash/api/src/model/util.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/util.ts:109]: Perfect, thanks!

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs:56]: Do you mean something that won't compile to Regex? Because that will already happen:
image

Also I stuck with panics because this isn't recoverable. We need a valid value for the domain validation URL

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (thehabbos007) [inline packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: I don't think this makes a big difference as the workspace would catch the invalid cases, but there's a check for not having `-` as the first symbol as well 
https://github.com/hashintel/hash/blob/1f85bd20b795939e7e1319d46a487d7a532844cc/packages/hash/api/src/model/knowledge/account.fields.ts#L46

And I assume `\w` includes underscores too, right?

Comment (thehabbos007) [inline packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: Though I think that having `|` is actually wrong here, it would match against the character which isn't valid in the workspace shortname validation

**Review by thehabbos007 — CHANGES_REQUESTED** [review]
One more thing, may be blocking

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: It's a good catch but I'm actively not trying to validate a shortname here, just to be able to capture an existing one.

We won't be creating shortnames based on these URLs, so I'm going to leave as is :)

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: Oops that's a good spot!
```suggestion
        default_value_t = Regex::new(r"http://localhost:3000/@(?P<shortname>(?:\w|-)+)/types/(?P<kind>(?:data-type)|(?:property-type)|(?:entity-type)|(?:link-type))/(?:\w|-)+/").unwrap(),
```

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: Although on that note I wonder if 
`[\w-]+` is better than `(?:\w|-)+`

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs:56]: This is an invalid regex and will fail when parsing the `Regex`. In this function you already have a valid regex, but it's not guaranteed that it fulfills our needs.

As this isn't recoverable and will not occur in runtime is good enough I guess, however, the error when passing an incomplete regex is not as nice as with `error-stack`:
```
$ cargo run -- --allowed-url-domain 'http://localhost:3000'

thread 'main' panicked at 'shortname capture group was missing', lib/graph/src/ontology/domain_validator.rs:41:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: packages/hash/integration/src/tests/util.ts
Comment (thehabbos007) [reply packages/hash/integration/src/tests/util.ts:99]: I think this is fine for now. A possibility would be to get around kratos in test-mode, but that's just more code to maintain than just randomly generating these user shortnames.

We could maybe look at allowing test-mode transactions in the Graph API (very basic Start transaction, roll back transaction APIs for testing only). But again, lots of code to maintain to make clean-up easier. I don't have good ideas for better test envs :/

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs:56]: This is only a nit and I don't want to block this PR on that, especially as we will revisit error handling very soon.

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: I think we don't want the `(?:\w|-)` grouped like this. It would create a non-capture group for each `\w`, wouldn't it? Perhaps `(?:[\w-]+)` would be better

File: packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/ontology/domain_validator.rs:56]: Okay so you were trying to ask if we should return errors instead of `expect` for the missing named capture groups.
Considering we don't have a consolidated error strategy yet, and we haven't discussed how to handle unrecoverable errors, I'm not going to change this to a result and an error here. Because as discussed, it's not recoverable so it should be a panic. If we want to look at using error-stack for panics that's a separate discussion I think we should have outside of this thread.

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: Is that a problem though? It's a _non_-capture group

regex101.com also isn't showing anything strange happening here but I'm not suggesting we can rely on it.
image

I'll probably swap to `[\w-]` anyway but I find that syntax a lot less clear because you don't have the explicit _or_

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]:
```suggestion
        default_value_t = Regex::new(r"http://localhost:3000/@(?P<shortname>[\w-]+)/types/(?P<kind>(?:data-type)|(?:property-type)|(?:entity-type)|(?:link-type))/[\w-]+/").unwrap(),
```

image

Comment (thehabbos007) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:39]: Ah no worries then. Up to you, either works for me if it doesn't mean it's wrong :)

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs]: I tried to google about efficiency and didn't get a clear answer but it is possible that the character classes are more efficient.

File: packages/graph/hash_graph/bin/hash_graph/src/args.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/bin/hash_graph/src/args.rs:40]: I think the frontend is running on `:3000` regardless of the `FRONTEND_DOMAIN` in dev mode (it just uses default next.js port on `yarn dev`).

**Review by TimDiekmann — DISMISSED** [review]
LGTM 🎉

**Review by thehabbos007 — APPROVED** [review]
Great job Alfie! Let's get it in :smile:

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**
